### ISSUE TITLE
Inconsistent behavior for report of IDE0067(DisposeObjectsBeforeLosingScope)

### ISSUE THREAD

**Issue by yyjdelete** [dotnet/roslyn#36498]
**Version Used**: 
`3.2.0-beta2-19303-01 (c9689b7a)`

**Steps to Reproduce**:
See IDE0067 is reported in some case but not for others.
```cs
using System.Collections.Generic;
using System.Threading;

public class Class1
{
    readonly List<CancellationTokenSource> list1 = new List<CancellationTokenSource>();
    readonly Stack<CancellationTokenSource> list2 = new Stack<CancellationTokenSource>();

    public void Test001()
    {
        list1.Add(new CancellationTokenSource());//1:not reported
        list2.Push(new CancellationTokenSource());//2:reported
        var cts = new CancellationTokenSource();//3:reported
        CancellationTokenRegistration cr = cts.Token.Register(() => { });//4:not reported
        //cr.Dispose();//CancellationTokenRegistration is IDisposable
    }
}
```
**Expected Behavior**:
1 and 2 should both be reported or not;
3 is OK;
4 should be reported.

**Actual Behavior**:
IDE0067 is reported for 2,3 but not for 1,4

**Comment by mavasani** [comment]
This is primarily because we have special cased semantics of few well known methods, but not all.

1. Add: Adding an object to a collection means it escapes analysis as the underlying collection might outlive the method scope. We are conservative and bail out.
2. Push: This is not special cased a well known method, so is equivalent to passing an object to a regular method invocation which does not change the dispose ownership of the passed object.
3. Seems okay
4. We do not treat disposable objects returned by all APIs as something that must be disposed by the caller, the dispose ownership is not really well defined and is specific to each API. We take the conservative approach of not tracking/flagging these objects, unless the method being invoked is in the same assembly for which we can do interprocedural analysis, i.e. if for case 4, Register was a private method returning a newly instantiated `CancellationTokenRegistration`, then we would flag it as we know the method is transferring dispose ownership to the caller.

We can probably special case more well known collection methods such as 2. to avoid these false positives.

**Comment by mavasani** [comment]
Discussed this offline with @jinujoseph and we are going to make the analysis conservative by default and assume that disposable object passed as argument to any external non-analyzed method is considered escaped. This should address false positives from case 2.

**Comment by mavasani** [comment]
Also reported [here](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/936137). https://github.com/dotnet/roslyn/pull/36674 resolves both these issues.
